CMake/coverage.cmake

```diff
@@ -35,8 +35,46 @@ function(add_converage module)
         add_custom_command(TARGET ${module}-ccov POST_BUILD
             COMMENT "Open ${module}-llvm-cov/index.html in your browser to view the coverage report."
         )
+    elseif("${CMAKE_C_COMPILER_ID}" MATCHES "GNU" OR "${CMAKE_CXX_COMPILER_ID}" MATCHES "GNU")
+        message("[Coverage] Building with Gcc Code Coverage Tools")
+
+        find_program(GCOV_PATH gcov)
+        IF(NOT GCOV_PATH)
+	       MESSAGE(FATAL_ERROR "gcov not found! Aborting...")
+        ENDIF() # NOT GCOV_PATH
+        find_program(GCOVR_PATH gcovr)
+        IF(NOT GCOVR_PATH)
+	       MESSAGE(FATAL_ERROR "gcovr not found! Aborting...")
+        ENDIF() # NOT GCOVR_PATH
+
+        set(COV_OUTPUT_PATH ${module}-gcc-cov)
+        target_compile_options(${module} PRIVATE -fprofile-arcs -ftest-coverage -fPIC )
+        target_link_options(${module} PRIVATE -fprofile-arcs -ftest-coverage -fPIC )
+        # this target is used for CI as well 
+        # generate the summary out.xml will send to github action to generate markdown, we can paste it to comments or readme
+        ADD_CUSTOM_TARGET(${module}-ccov
+            COMMAND ${module}  ${TEST_PARAMETER}
+            COMMAND rm -rf ${COV_OUTPUT_PATH}
+            COMMAND mkdir ${COV_OUTPUT_PATH}
+            COMMAND gcovr -r ${CMAKE_SOURCE_DIR} --object-directory=${PROJECT_BINARY_DIR} -x ${COV_OUTPUT_PATH}/out.xml --exclude='${CMAKE_SOURCE_DIR}/unittests/' --exclude='${PROJECT_BINARY_DIR}/'
+            COMMAND gcovr -r ${CMAKE_SOURCE_DIR} --object-directory=${PROJECT_BINARY_DIR} --html ${COV_OUTPUT_PATH}/report.html --exclude='${CMAKE_SOURCE_DIR}/unittests/' --exclude='${PROJECT_BINARY_DIR}/'
+            WORKING_DIRECTORY ${PROJECT_BINARY_DIR}
+            COMMENT "Running gcovr to produce Cobertura code coverage report."
+        )
+
+        # generate the detail report
+        ADD_CUSTOM_TARGET(${module}-ccov-report
+        COMMAND ${module}  ${TEST_PARAMETER}
+            COMMAND rm -rf ${COV_OUTPUT_PATH}
+            COMMAND mkdir ${COV_OUTPUT_PATH}
+            COMMAND gcovr -r ${CMAKE_SOURCE_DIR} --object-directory=${PROJECT_BINARY_DIR}  --html-details ${COV_OUTPUT_PATH}/index.html --exclude='${CMAKE_SOURCE_DIR}/unittests/' --exclude='${PROJECT_BINARY_DIR}/'
+            WORKING_DIRECTORY ${PROJECT_BINARY_DIR}
+            COMMENT "Running gcovr to produce Cobertura code coverage report."
+        )
+        add_custom_command(TARGET ${module}-ccov-report POST_BUILD
+        COMMENT "Open ${COV_OUTPUT_PATH}/index.html in your browser to view the coverage report."
```

Comment:
Could use a tab or some spaces to make it look more in line with the other formatting of this file. While this happens automatically for C++ files with `clang-format`, unfortunately, we have to maintain it ourselves in other files :-(

Reply:
will update these format issue. :). I just compose this from other people's code, that's why u see a lot of format mismatch~